tree_unique_args/src/ir.rs

```diff
@@ -0,0 +1,223 @@
+        }
+    }
+
+    pub(crate) fn name(&self) -> &'static str {
+        self.to_sort().name()
+    }
+}
+
+#[derive(Clone, Copy, EnumIter)]
+pub(crate) enum Constructor {
+    Num,
+    Boolean,
+    UnitExpr,
+    Add,
+    Sub,
+    Mul,
+    LessThan,
+    And,
+    Or,
+    Not,
+    Get,
+    Print,
+    Read,
+    Write,
+    All,
+    Switch,
+    Loop,
+    Body,
+    Arg,
+    Call,
+    Cons,
+    Nil,
+}
+
+// The constructor fields must purposes such that this is maintained:
+// - A ctor has one or more CapturedExpr fields iff it has exactly one
+//   CapturingId field. The CapturingId field corresponds to the context of the
+//   CapturedExpr field(s).
+//   * Note that this applies to body/loop ids, but not the id in an arg.
+//   * Note also that a call's function reference has purpose Static
+// Invariants of a valid term in the IR:
+// - A ReferencingId must match the nearest enclosing BindingId
+// - It must typecheck (see typechecker in interpreter.rs).
+pub(crate) enum Purpose {
+    Static(Sort), // some int, bool, order that parameterizes constructor
+    CapturingId,
+    ReferencingId,
+    SubExpr,      // subexpression, e.g. Add's summand
+    SubListExpr,  // sublistexpr, e.g. Switch's branch lsit
+    CapturedExpr, // a body's outputs
+}
+
+impl Purpose {
+    pub(crate) fn to_sort(&self) -> Sort {
+        match self {
+            Purpose::CapturingId => Sort::I64,
+            Purpose::ReferencingId => Sort::I64,
+            Purpose::SubExpr => Sort::Expr,
+            Purpose::CapturedExpr => Sort::Expr,
+            Purpose::SubListExpr => Sort::ListExpr,
+            Purpose::Static(sort) => *sort,
+        }
+    }
+}
+
+pub(crate) struct Field {
+    pub purpose: Purpose,
+    pub name: &'static str,
+}
+
+impl Field {
+    pub(crate) fn sort(&self) -> Sort {
+        self.purpose.to_sort()
+    }
+}
+
+impl Constructor {
+    pub(crate) fn name(&self) -> &'static str {
+        match self {
+            Constructor::Num => "Num",
+            Constructor::Boolean => "Boolean",
+            Constructor::UnitExpr => "UnitExpr",
+            Constructor::Add => "Add",
+            Constructor::Sub => "Sub",
+            Constructor::Mul => "Mul",
+            Constructor::LessThan => "LessThan",
+            Constructor::And => "And",
+            Constructor::Or => "Or",
+            Constructor::Not => "Not",
+            Constructor::Get => "Get",
+            Constructor::Print => "Print",
+            Constructor::Read => "Read",
+            Constructor::Write => "Write",
+            Constructor::All => "All",
+            Constructor::Switch => "Switch",
+            Constructor::Loop => "Loop",
+            Constructor::Body => "Body",
+            Constructor::Arg => "Arg",
+            Constructor::Call => "Call",
+            Constructor::Cons => "Cons",
+            Constructor::Nil => "Nil",
+        }
+    }
+
+    pub(crate) fn fields(&self) -> Vec<Field> {
```

Comment:
I think fields should probably have an underscore or something to help reduce chance of aliasing

Reply:
What might it alias with?

Reply:
Other variables in your pattern.
For example, if I use my own `x` inside a pattern as well as use your fields

Reply:
Ah, I misread it as saying the method `fields` itself would alias 🤦.

Rather than "default var names," I think field names should be thought of as a more human-readable alternative to indices. They allow us to do [this equality check](https://github.com/egraphs-good/eggcc/blob/63894d78dc04e0f9c0a38cc486bbf221507e80b1/tree_unique_args/src/cf_invariant_code_motion.rs#L17) soundly, and yes, generate variable names better than `x0`, `x1`, `x2`.

I am in favor of not including an underscore since users generating rules can easily append disambiguation themselves if necessary.

Reply:
This is a pretty obvious footgun, we should try to mitigate it

The generated code can be a little messier

Reply:
Discussed on Slack - added a `.var()` method that prepends an underscore